src/Npgsql/NpgsqlActivitySource.cs

```diff
@@ -20,19 +21,56 @@ static NpgsqlActivitySource()
 
     internal static bool IsEnabled => Source.HasListeners();
 
-    internal static Activity? CommandStart(NpgsqlConnector connector, string sql)
+    internal static Activity? CommandStart(NpgsqlConnector connector, string commandText, CommandType commandType, bool storedProcCompatMode)
     {
         var settings = connector.Settings;
-        var activity = Source.StartActivity(settings.Database!, ActivityKind.Client);
+
+        var dbName = settings.Database ?? connector.InferredUserName;
```

Comment:
This feels like something that could make sense on the connector (I could imagine other uses for it), though we can also do that when the need arises.

Reply:
You mean something like `ActualDbName` which is guaranteed to never be null?
I agree that the will probably be helpful, especially in any kind of logging scenario.
This is actually a concrete case of the abstract problem I was trying to describe when doing initial investigations for https://github.com/npgsql/npgsql/issues/4462. We are inferring different things in different places from different constant or runtime defaults (there's also the PG... environment variables) and sometimes cache the result (again in different places) and sometimes don't.
What I envision is some central immutable structure (maybe on the connector, maybe on the connection maybe on the data source) where we store all the *actual* values we use.

Reply:
Yeah, exactly. And I think you're right that it makes even more sense on NpgsqlDataSource (though note that MultiHostDataSource wouldn't be able to have it, it could throw instead).

Reply:
> though note that MultiHostDataSource wouldn't be able to have it

Thinking about this, it could probably have most of it (obviously not the host/port) since it has multiples but inferred user name and database name would work. Maybe even an abstract structure with two implementations (one for single host and one for multiple hosts) that only differ where it makes sense.
There's also even some inferred (rather resolved) information that can only live on the individual connector while it is open (e. g. the resolved IP endpoint) but that might still be of use for logging purposes.